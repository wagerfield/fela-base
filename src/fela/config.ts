import { IConfig as FelaConfig } from "fela"
import monolithic from "fela-monolithic"
import extendPlugin from "fela-plugin-extend"
import embeddedPlugin from "fela-plugin-embedded"
import fallbackPlugin from "fela-plugin-fallback-value"
import pseudoPlugin from "fela-plugin-friendly-pseudo-class"
import placeholderPlugin from "fela-plugin-placeholder-prefixer"
import prefixerPlugin from "fela-plugin-prefixer"
import lvhaPlugin from "fela-plugin-lvha"
import unitPlugin, { Unit, UnitPerProperty } from "fela-plugin-unit"
import { Aliases, Breakpoints, CustomProps, RuleDefaults } from "../types"
import { setRuleDefaults } from "../rules/defaults"
import customPropsPlugin from "./custom-props"
import namedKeysPlugin from "./named-keys"

export interface UnitOptions {
  unitMap: UnitPerProperty
  defaultUnit: Unit
}

export interface ConfigOptions {
  useMonolithicClasses?: boolean
  usePrettySelectors?: boolean
  ruleDefaults?: RuleDefaults
  customProps?: CustomProps
  breakpoints?: Breakpoints
  aliases?: Aliases
  units?: UnitOptions
}

export default ({
  useMonolithicClasses = false,
  usePrettySelectors = false,
  ruleDefaults,
  customProps,
  breakpoints,
  aliases,
  units
}: ConfigOptions = {}): FelaConfig => {
  // Set rule defaults
  if (ruleDefaults) setRuleDefaults(ruleDefaults)

  // Add monolithic enhancer
  const enhancers = useMonolithicClasses
    ? [monolithic({ prettySelectors: usePrettySelectors })]
    : []

  // Unit plugin values
  const defaultUnit = (units && units.defaultUnit) || "px"
  const unitMap = (units && units.unitMap) || {}

  // Add plugins: order matters
  const plugins = [
    extendPlugin(),
    pseudoPlugin(),
    namedKeysPlugin({ aliases, breakpoints }),
    customPropsPlugin(customProps),
    placeholderPlugin(),
    embeddedPlugin(),
    prefixerPlugin(),
    fallbackPlugin(),
    lvhaPlugin(),
    unitPlugin(defaultUnit, unitMap)
  ]

  // Return config
  return { enhancers, plugins }
}
