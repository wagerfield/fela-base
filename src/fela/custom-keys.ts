import { TPlugin as FelaPlugin } from "fela"
import customKeys, {
  Config as CustomKeys
} from "fela-plugin-custom-property"
import { ALL } from "../core/edges"
import { isObject } from "../core/utils"
import ellipsis from "../rules/ellipsis/ellipsis"
import border from "../rules/border/border"
import edges from "../rules/edges/edges"

export {
  Config as CustomKeys
} from "fela-plugin-custom-property"

export default (options: CustomKeys = {}): FelaPlugin =>
  customKeys({
    border: (value) =>
      isObject(value) ? border(value) : { border: value },

    ellipsis: (value) => (value ? ellipsis() : {}),

    fontFeatures: (value) => ({
      "-webkit-font-feature-settings": value,
      "font-feature-settings": value
    }),

    fontSmoothing: (value) => ({
      "-webkit-font-smoothing": value,
      "-moz-osx-font-smoothing": "grayscale"
    }),

    margin: (value) =>
      edges({
        prefix: "margin",
        edges: ALL,
        value
      }),

    padding: (value) =>
      edges({
        prefix: "padding",
        edges: ALL,
        value
      }),

    scrolling: (value) => ({
      "-webkit-overflow-scrolling": value
    }),

    size: (value) => ({
      width: value,
      height: value
    }),

    tapColor: (value) => ({
      "-webkit-tap-highlight-color": value
    }),

    ...options
  })
