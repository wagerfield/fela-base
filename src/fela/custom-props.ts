import { TPlugin as FelaPlugin } from "fela"
import customKeys from "fela-plugin-custom-property"
import { border, edge, ellipsis, outline } from "../rules"
import { CustomProps, Rule } from "../types"
import { isObject } from "../core/utils"
import { ALL } from "../core/edges"

const customProp = (rule: Rule, key: string) => (value: any) =>
  isObject(value) ? rule(value) : { [key]: value }

export default (props?: CustomProps): FelaPlugin =>
  customKeys({
    border: customProp(border, "border"),

    outline: customProp(outline, "outline"),

    ellipsis: (value) => (value ? ellipsis() : {}),

    fontFeatures: (value) => ({
      "-webkit-font-feature-settings": value,
      "font-feature-settings": value
    }),

    fontSmoothing: (value) => ({
      "-webkit-font-smoothing": value,
      "-moz-osx-font-smoothing": "grayscale"
    }),

    margin: (value) =>
      edge({
        prefix: "margin",
        edge: ALL,
        value
      }),

    padding: (value) =>
      edge({
        prefix: "padding",
        edge: ALL,
        value
      }),

    scrolling: (value) => ({
      "-webkit-overflow-scrolling": value
    }),

    size: (value) => ({
      width: value,
      height: value
    }),

    tapColor: (value) => ({
      "-webkit-tap-highlight-color": value
    }),

    ...props
  })
