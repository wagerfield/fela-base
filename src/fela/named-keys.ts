import { objectReduce } from "fast-loops"
import { TPlugin as FelaPlugin } from "fela"
import namedKeys from "fela-plugin-named-keys"

import { kebabToCamel } from "../core/utils"

export interface Breakpoints {
  [name: string]: number
}

export interface Aliases {
  [name: string]: string
}

export interface NamedKeysOptions {
  aliases?: Aliases
  breakpoints?: Breakpoints
}

export const buildMediaQuery = (
  value: number,
  below: boolean
) => {
  const prefix = below ? "max" : "min"
  const offset = below ? value : value + 1
  return `@media (${prefix}-width: ${offset}px)`
}

export const buildMediaQueries = (
  breakpoints: Breakpoints
) => {
  const acc: Aliases = {}
  return objectReduce(
    breakpoints,
    (result, value, key) => {
      const belowKey = kebabToCamel(`below-${key}`)
      const aboveKey = kebabToCamel(`above-${key}`)
      result[belowKey] = buildMediaQuery(value, true)
      result[aboveKey] = buildMediaQuery(value, false)
      return result
    },
    acc
  )
}

export default ({
  aliases = {},
  breakpoints = {}
}: NamedKeysOptions = {}): FelaPlugin =>
  namedKeys({
    placeholder: "::placeholder",
    ...buildMediaQueries(breakpoints),
    ...aliases
  })
