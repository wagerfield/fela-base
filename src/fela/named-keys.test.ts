import { createRenderer, IStyle } from "fela"
import namedKeys, {
  buildMediaQuery,
  buildMediaQueries,
  NamedKeysOptions
} from "./named-keys"

interface CustomStyle extends IStyle {
  aboveMobile?: object
  belowMobile?: object
  aboveTablet?: object
  belowTablet?: object
  placeholder?: object
  supportsFlex?: object
}

const snapshotStyle = (name?: string) => (style) => {
  expect(style).toMatchSnapshot(name)
  return style
}

const testPlugin = (style: CustomStyle, options?: NamedKeysOptions) => {
  createRenderer({
    plugins: [namedKeys(options), snapshotStyle()]
  }).renderRule(() => style, {})
}

describe("buildMediaQuery(value, below)", () => {
  it("returns expected snapshots", () => {
    expect(buildMediaQuery(320, true)).toMatchSnapshot()
    expect(buildMediaQuery(320, false)).toMatchSnapshot()
  })
})

describe("buildMediaQueries(breakpoints)", () => {
  it("returns expected snapshots", () => {
    expect(
      buildMediaQueries({
        mobile: 320,
        tablet: 720
      })
    ).toMatchSnapshot()
  })
})

describe("namedKeys(options)", () => {
  it("returns a function", () => {
    expect(namedKeys()).toEqual(expect.any(Function))
  })
  it("renames placeholder alias", () => {
    testPlugin({
      placeholder: {
        color: "red"
      }
    })
  })
  it("renames media query aliases", () => {
    testPlugin(
      {
        aboveMobile: {
          color: "red"
        },
        belowTablet: {
          color: "red"
        }
      },
      {
        breakpoints: {
          mobile: 320,
          tablet: 720
        }
      }
    )
  })
  it("renames custom aliases", () => {
    testPlugin(
      {
        supportsFlex: {
          color: "red"
        }
      },
      {
        aliases: {
          supportsFlex: "@supports (display: flex)"
        }
      }
    )
  })
})
