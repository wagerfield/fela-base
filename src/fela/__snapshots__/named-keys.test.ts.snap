// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildMediaQueries(breakpoints) returns expected snapshots 1`] = `
Object {
  "aboveMobile": "@media (min-width: 321px)",
  "aboveTablet": "@media (min-width: 721px)",
  "belowMobile": "@media (max-width: 320px)",
  "belowTablet": "@media (max-width: 720px)",
}
`;

exports[`buildMediaQuery(value, below) returns expected snapshots 1`] = `"@media (max-width: 320px)"`;

exports[`buildMediaQuery(value, below) returns expected snapshots 2`] = `"@media (min-width: 321px)"`;

exports[`namedKeys(options) renames custom aliases 1`] = `
Object {
  "@supports (display: flex)": Object {
    "color": "red",
  },
}
`;

exports[`namedKeys(options) renames media query aliases 1`] = `
Object {
  "@media (max-width: 720px)": Object {
    "color": "red",
  },
  "@media (min-width: 321px)": Object {
    "color": "red",
  },
}
`;

exports[`namedKeys(options) renames placeholder alias 1`] = `
Object {
  "::placeholder": Object {
    "color": "red",
  },
}
`;
